docs/design/coreclr/botr/vectors-and-intrinsics.md

```diff
@@ -0,0 +1,189 @@
+- Ahead of time generated code which could take advantage of more advanced hardware support experiences a performance penalty until rejitted. (If a customer chooses to disable tiered compilation, then customer code may always run slowly).
+
+#### Code review rules for code written in System.Private.CoreLib.dll
+- Any use of a platform intrinsic in the codebase MUST be wrapped with a call to the associated IsSupported property. This wrapping MUST be done within the same function that uses the hardware intrinsic, and MUST NOT be in a wrapper function unless it is one of the intrinsics that are enabled by default for crossgen compilation of System.Private.CoreLib (See list above in the implementation rules section). 
+- Within a single function that uses platform intrinsics, it must behave identically regardless of whether IsSupported returns true or not. This rule is required as code inside of an IsSupported check that calls a helper function cannot assume that the helper function will itself see its use of the same IsSupported check return true. This is due to the impact of tiered compilation on code execution within the process. 
+- Excessive use of intrinsics may cause startup performance problems due to additional jitting, or may not achieve desired performance characteristics due to suboptimal codegen.
+
+ACCEPTABLE Code
+```csharp
+using System.Runtime.Intrinsics.X86;
+
+public class BitOperations
+{
+    public static int PopCount(uint value)
+    {
+        if (Avx2.IsSupported)
+        {
+            Some series of Avx2 instructions that performs the popcount operation.
+        }
+        else
+            return FallbackPath(input);
+    }
+
+    private static int FallbackPath(uint)
+    {
+        const uint c1 = 0x_55555555u;
+        const uint c2 = 0x_33333333u;
+        const uint c3 = 0x_0F0F0F0Fu;
+        const uint c4 = 0x_01010101u;
+
+        value -= (value >> 1) & c1;
+        value = (value & c2) + ((value >> 2) & c2);
+        value = (((value + (value >> 4)) & c3) * c4) >> 24;
+
+        return (int)value;
+    }
+}
+```
+
+UNACCEPTABLE code
+```csharp
+using System.Runtime.Intrinsics.X86;
+
+public class BitOperations
+{
+    public static int PopCount(uint value)
+    {
+        if (Avx2.IsSupported)
+            return UseAvx2(value);
+        else
+            return FallbackPath(input);
+    }
+
+    private static int FallbackPath(uint)
+    {
+        const uint c1 = 0x_55555555u;
+        const uint c2 = 0x_33333333u;
+        const uint c3 = 0x_0F0F0F0Fu;
+        const uint c4 = 0x_01010101u;
+
+        value -= (value >> 1) & c1;
+        value = (value & c2) + ((value >> 2) & c2);
+        value = (((value + (value >> 4)) & c3) * c4) >> 24;
+
+        return (int)value;
+    }
+
+    private static int UseAvx2(uint value)
+    {
+        // THIS IS A BUG!!!!!
```

Comment:
Is it worth documenting that this is a bug because `PopCount` might have been rejitted to know `Avx2.IsSupported` but `UseAvx2` might have not?

Reply:
Likewise, is this "ok" to do if `UseAvx2` is marked `AggressiveOptimization`?

Reply:
I wouldn't think that this would be ok even if it is marked `AggressiveOptimization` - I don't think we should *ever* rely for correctness on the behavior of optimization.

Reply:
I agree with Carol here. At some point we may change the policy on `AggressiveOptimization`, and that shouldn't break the correctness of applications. We do parse another attribute, the Ssytem.Runtime.BypassReadyToRunAttribute, which can be used to simply declare that a method shouldn't be ready to runned. This attribute isn't actually in the BCL, the application must define it itself, but it can be used to safely forbid a method from being precompiled. I've updated the document, and added a comment that this pattern IS acceptable when applied to a subset of the x86 and x64 instruction sets, as we have special treatment for those instruction sets in the aot compiler when compiling CoreLib.